Endianness.md

```diff
@@ -13,7 +13,7 @@ Some compilers or system headers provide macros to determine endianness, but the
 
 * Avoid endianness whenever possible.
 * Use endian-neutral code (see example below.)
-* Use existing conversion functions, such as `htons()` to convert between network byte-order and the native byte-order of the processor.
+* Use existing conversion functions, such as `htons()` and `htonl()` to convert from network to host byte-order and `bswap_16()`, `bswap_32()`, and `bswap_64()` to covert to the native byte-order of the processor.
```

Comment:
This is incorrect.

1. The `hton` class of functions converts from `host` to `network` byte order, where as `ntoh` from `network` to `host`, thus the name. Despite they're probably doing the same thing, it's separate functions after all.

2. I don't think `bswap_{16,32,64}` are standard functions. If they were added in newer C and C++ standard, that better to be mentioned. Also, they do not convert to `native` order, they just swap bytes, thus the name.

Reply:
> The hton class of functions converts from host to network byte order, where as ntoh from network to host, thus the name. Despite they're probably doing the same thing, it's separate functions after all.

You're correct. Despite I've specified correct term in the commit message, I absolutely missed this. Thanks for the correction.

> I don't think bswap_{16,32,64} are standard functions. If they were added in newer C and C++ standard, that better to be mentioned.

They indeed aren't standard. When compiling with GCC/Clang, they just get inlined to builtins, (e.g. `__builtin_bswap32(...)`, which compiler turns into a `bswap` or similar instruction, assuming the CPU does support it or fall-back to native implementation. These `__builtins` (swap), were added in GCC 4.x, (`__builtin_bswap16()` was a bit late), but from GCC 5, this isn't a problem at all. I'm not exactly sure when LLVM added these, but I assume, it was a long time ago. I'm not sure what MSVC does in this case, but I think, I'll rather
specify the `__builtin_*` functions over these wrappers, as they only target GNU systems, and so system like FreeBSD, wouldn't able to understand these. 

> Also, they do not convert to native order, they just swap bytes, thus the name.

I'm not sure if you mean by the name of these function, but just "swapping" can sound completely different. 

Here's the glibc implementation of  `htons()`
```c
uint16_t
htons (uint16_t x)
{
#if BYTE_ORDER== BIG_ENDIAN
  return x;
#elif BYTE_ORDER == LITTLE_ENDIAN
  return __bswap_16(x);
#else
# error "What kind of system is this?"
#endif
}
```

As you can see, if the source was compiled on a big-endian platform, it will just return what it got, and on little-endian platform, it will do the byteswap.

Basically, `bswap*` will be done regardless your CPU endian support, whereas `htons()`, in this case, will only work if host is also little-endian.

Reply:
> They indeed aren't standard. When compiling with GCC/Clang, they just get inlined to builtins, (e.g. 

C++23 has added [`std::byteswap`](https://en.cppreference.com/w/cpp/numeric/byteswap).

Reply:
> > They indeed aren't standard. When compiling with GCC/Clang, they just get inlined to builtins, (e.g.
> 
> C++23 has added [`std::byteswap`](https://en.cppreference.com/w/cpp/numeric/byteswap).

Yes, but that's just for C++ only. For C, you either have to use bswap_* (for GNU systems) or bswap* (for BSD systems) or __builtin_bswap*.

Reply:
The main purpose of the predef pages is to list pre-defined macros. The Endianness page went a bit further by mentioning an overall approach to how endianess may be detected. If we get into the details of builtin functions then we get further out of scope.

I would rather delete the mentioning of `htons` and simply state "Use existing conversion functions to convert between network byte-order and the native byte-order of the processor."

Reply:
> The main purpose of the predef pages is to list pre-defined macros. The Endianness page went a bit further by mentioning an overall approach to how endianess may be detected. If we get into the details of builtin functions then we get further out of scope.
> 
> I would rather delete the mentioning of `htons` and simply state "Use existing conversion functions to convert between network byte-order and the native byte-order of the processor."

That's actually will cleanup a few things, but it's upto @ a1batross (if they want to) :)

Reply:
I'm absolutely fine with that.

Reply:
Thanks! I've updated it.

Reply:
It should be okay now.